### PR TITLE
add wsv restore mode - sync blockstore

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
add --drop_state param, param --reuse_state is default

Signed-off-by: Alexey-N-Chernyshov <alexey.n.chernyshov@gmail.com>

### CODE REVIEW

**Review by kamilsa — APPROVED** [review]

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 1]
```diff
@@ -160,3 +161,16 @@ FlatFile::FlatFile(std::string path,
...
+void FlatFile::reload() const {
+  available_blocks_.clear();
+  for (auto it = boost::filesystem::directory_iterator{dump_dir_};
+       it != boost::filesystem::directory_iterator{};
+       ++it) {
+    if (auto id = FlatFile::name_to_id(it->path().filename().string())) {
+      available_blocks_.insert(*id);
+    } else {
+      boost::filesystem::remove(it->path());
+    }
```
Comment (iceseer) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp:174]: You don't need to clear this cache and iterate through the whole block store on each call. You know the size of cache and you expect the exact file with determined name.

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 2]
```diff
@@ -139,6 +139,7 @@ std::string FlatFile::directory() const {
 }
 
 Identifier FlatFile::last_id() const {
+  reload();
   return (available_blocks_.empty()) ? 0 : *available_blocks_.rbegin();
```
Comment (iceseer) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp:143]: I expect this method to be constant and very light. To make multiple FS calls is not a good decision. This mode must be disconnectable. The best way I see is an option for FlatFile to skip write blocks. And independent mechanics, that scans the directory for new and not locked expected files, that can be read.

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 3]
```diff
@@ -160,3 +161,16 @@ FlatFile::FlatFile(std::string path,
...
       available_blocks_(std::move(existing_files)),
       log_{std::move(log)} {}
+
+void FlatFile::reload() const {
+  available_blocks_.clear();
+  for (auto it = boost::filesystem::directory_iterator{dump_dir_};
+       it != boost::filesystem::directory_iterator{};
+       ++it) {
+    if (auto id = FlatFile::name_to_id(it->path().filename().string())) {
+      available_blocks_.insert(*id);
```
Comment (iceseer) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp:171]: We can't add this file for processing, until we are sure, that the file is released by other process.

File: irohad/ametsuchi/impl/wsv_restorer_impl.cpp [hunk 4]
```diff
@@ -156,83 +166,107 @@ namespace iroha::ametsuchi {
...
+          return reindexBlocks(storage,
+                               mutable_storage,
+                               *block_query,
+                               *interface_validator_,
+                               *proto_validator_,
+                               *validator_,
+                               wsv_ledger_height + 1,
+                               last_block_in_storage);
+        };
+        while (wait_for_new_blocks) {
```
Comment (iceseer) [inline irohad/ametsuchi/impl/wsv_restorer_impl.cpp:257]: There is no ```res``` check.

**Review by iceseer — CHANGES_REQUESTED** [review]

**Review by iceseer — APPROVED** [review]

**Pull request merged**